### ISSUE TITLE
Security

### ISSUE THREAD

**Issue by fwsGonzo** [libriscv/godot-sandbox#64]
This is the big security issue... list. Things that I believe are important to do, and to avoid.

## What is Godot Sandbox

First off, let's define the Godot Sandbox. What will it be? It's going to be a memory- and system- safe protection layer that prevents anyone from reaching the outside system. At most, people will be able to modify and trash the game. That's fine, and we can call that a trashing mod. What's not OK is executing code outside of Godot, modifying the file system or in any way reading out user information.

## What is GDScript

Today, people write GDScript to create games and occasionally to let people modify a game. GDScript can be evaluated and it is the most integrated solution in Godot. It is not, however, safe. Try pasting `-9223372036854775808 / -1` anywhere in a GDScript window:
```
handle_crash: Program crashed with signal 8
```

So, GDScript is not a safe scripting language by any stretch of the imagination. Instead we have to reduce the surface area greatly, and do what we can to **completely avoid** GDScript interaction.

## The current implementation

The current implementation is most like not safe, but it has a few architectural ideas that may turn out to be safe:
- Variants are always translated in and out of the guest
     - Except callables, which are allow-listed for the duration of the call. This can only happen by someone passing a callable to the guest, which implies that it is OK to call it.
- Objects and Nodes must be seen in order to be operable
    - After a call ends, all knowledge of nodes and objects evaporate
    - All deletion/freeing is deferred to the next frame
    - It is possible to make arbitrary calls on Nodes and Objects, which is a great thing to support, but breaks the sandbox

### Calling on Objects and Nodes

Calls are great, but as noted before, they break the sandbox. There are probably:
- Ways to create call loops
    - The sandbox will actively prevent recursive calls reaching depth >= 8
    - You may still be able to get Godot to call loop
- Calls can call functions like `free` and `remove` that are dangerous
- Calls may be able to sneakily access global objects like `OS`

### Creating malicious trees

The sandbox gives the user access to node manipulation, including adding children. As such, one can create a hierarchy of nodes of ones own choosing. It is very likely possible to create a tree in such a way that an exploit can happen.

The sandbox places limits on objects and nodes that can be active, but ultimately the limit will likely continually be increased in order to not constrain modders. We may as well say that the limit will tend to what the game can bear.

### Execution timeouts

Execution timeout is possibly impossible to rein in. There are too many operations that can take a long time, and as such, trying to limit the sandbox to a certain time-frame without signal interruption is hopeless. Infinite loops are taken care of by instruction counting, and many costly system calls have penalties that increase the counter suitably. Nevertheless, there are many operations in Godot that modders must have access to, that would be able to keep the system hanging forever.

In short, we can only say that execution timeout is present, but not a guarantee. Nor is it always wanted. People will just close the game if it's not working as intended.

### Timing gadgets

It is likely easy to find timing gadgets, so I am not sure if lowering the granularity of the usual timers are going to do anything but make a side-channel attack demo take longer to make. The sandbox is supposed to be speculation-safe, but we cannot guarantee anything outside of it, and there is too much sandbox-reachable code in Godot. I suspect timing attacks will be possible, with enough time. This is, unfortunately, a topic that I have very little knowledge of, so take all this with a grain of salt. Should you, dear reader, have any knowledge, please let me know what you think.

**Comment by fwsGonzo** [comment]
The Godot Sandbox has recently gained a new restriction system as well as some hardening against self-tampering (sandbox trying to unrestrict itself through the APIs).
